### PR TITLE
torchscript compatible SRU init commit

### PR DESCRIPTION
**Pull request by taoleicn**
- [x] address docstring length
- [x] remove unused import
- [x] move unit tests 
- [x] resolve c++ test error in pytorch 1.3/1.4
- [x] support optional tensor input in cpp 
- [x] add unit test for loading and verifying model in c++
- [x] address flake8 / mypy errors
- [x] automate flake8 test

### CODE REVIEW

File: sru/csrc/main_test_cpp.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+
+#include <torch/script.h> // One-stop header.
+
+#include <iostream>
+#include <memory>
+
+
+int main(int argc, const char* argv[]) {
```
Comment (jwohlwend) [inline sru/csrc/main_test_cpp.cpp:8]: while this kind of test is useful, it would be good to have a test that checks that the output is equal when calling from python or c++. Not quite sure what the best way to do that is but seems important

File: sru/csrc/sru_cpu_impl.cpp [hunk 2]
```diff
@@ -255,7 +256,12 @@ inline float apply_activation(int64_t type, float x) {
...
 
-PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
-  m.def("cpu_forward", &cpu_forward, "SRU forward (CPU version)");
-  m.def("cpu_bi_forward", &cpu_bi_forward, "SRU bidirectional forward (CPU version)");
+TORCH_LIBRARY(sru_cpu, m) {
+  m.def("cpu_forward", &cpu_forward);
+  m.def("cpu_bi_forward", &cpu_bi_forward);
 }
+
+//PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
```
Comment (jwohlwend) [inline sru/csrc/sru_cpu_impl.cpp:264]: maybe remove commented code?

File: sru/modules.py [hunk 3]
```diff
@@ -3,230 +3,78 @@
...
-                 bidirectional=False,
-                 n_proj=0,
-                 use_tanh=0,
-                 highway_bias=0,
-                 has_skip_term=True,
-                 layer_norm=False,
-                 rescale=True,
-                 v1=False,
-                 custom_m=None):
+                 input_size: int,
```
Comment (jwohlwend) [inline sru/modules.py:30]: nice adding the typing :)

File: sru/ops.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+            h[t, :, di, :] = h_t
+        c_final.append(c_t.view(batch, d))
+
+    return h.view(length, batch, -1), torch.stack(c_final, dim=1).view(batch, -1)
+
+
+
+# some test code
+if __name__ == "__main__":
+    print(elementwise_recurrence_cpu)
```
Comment (hpasapp) [inline sru/ops.py:217]: can we move this into a pytest script please?

File: sru/modules.py [hunk 5]
```diff
@@ -3,230 +3,78 @@
...
+                 bidirectional: bool = False,
+                 n_proj: int = 0,
+                 use_tanh: bool = False,
+                 highway_bias: float = 0.0,
+                 has_skip_term: bool = True,
+                 layer_norm: bool = False,
+                 rescale: bool = True,
+                 v1: bool = False,
+                 custom_m: Optional[nn.Module] = None):
+        """Initialize the SRUCell module.
```
Comment (jwohlwend) [inline sru/modules.py:43]: In general it's good to keep docstrings under 72 chars for parsing tools like sphynx, just FYI :)

File: sru/ops.py (hunk 4 above)
Comment (hpasapp) [reply sru/ops.py:217]: (like: move this into a new file, eg `tests/sru/test_ops.py`, in a function like eg `def test_all()`, or similar.
- replace `print`s by `assert`s)

File: sru/modules.py [hunk 6]
```diff
@@ -659,3 +600,42 @@ def make_backward_compatible(self):
             self.input_to_hidden = None
             for cell in self.rnn_lst:
                 cell.custom_m = None
+
+# some test code
+if __name__ == "__main__":
```
Comment (hpasapp) [inline sru/modules.py:605]: ditto for this: can we move this into a test script in `tests/sru` please?

File: sru/modules.py [hunk 7]
```diff
@@ -659,3 +600,42 @@ def make_backward_compatible(self):
             self.input_to_hidden = None
             for cell in self.rnn_lst:
                 cell.custom_m = None
+
+# some test code
```
Comment (jwohlwend) [inline sru/modules.py:604]: maybe not the best spot for testing code?

**Comment by hpasapp** [comment]
flake8 shows the following lines are not used:
```
import os
import sys
from typing import Overload
import torch.nn.functional as F
...
 input_size, hidden_size = self.input_size, self.hidden_size
```
(last line is from line 206)

File: sru/__init__.py [hunk 8]
```diff
@@ -5,7 +5,4 @@
 in "Simple Recurrent Units for Highly Parallelizable Recurrence."
 """
 from .version import __version__
-try:
-    from .sru_functional import *
-except:
-    from sru_functional import *
+from .modules import *
```
Comment (jwohlwend) [inline sru/__init__.py:8]: in general i find it more robust to do: `from sru.modules import *`. Aside, seems like the try except was there for a reason, any reason to remove it?

File: sru/ops.py (hunk 4 above)
Comment (jwohlwend) [reply sru/ops.py:217]: +1

File: sru/modules.py (hunk 6 above)
Comment (jwohlwend) [reply sru/modules.py:605]: +1

File: sru/csrc/main_test_cpp.cpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (hpasapp) [inline sru/csrc/main_test_cpp.cpp:1]: can we consider moving this file into a folder with a name like `tests`, to keep it separate from the actual sru code please?

File: sru/csrc/sru_cpu_impl.cpp [hunk 10]
```diff
@@ -1,5 +1,6 @@
 
-#include <torch/torch.h>
+//#include <torch/torch.h>
```
Comment (hpasapp) [inline sru/csrc/sru_cpu_impl.cpp:2]: can we remove this dead code altogether please?

File: sru/csrc/sru_cpu_impl.cpp (hunk 2 above)
Comment (hpasapp) [inline sru/csrc/sru_cpu_impl.cpp:264]: can we remove this dead code please?

**Comment by jwohlwend** [comment]
As @hpasapp mentioned, tests should leverage pytest and run using the integrated github CI. I think it would also be good to 1) check that old checkpoints can be reloaded 2) that C++ and python result in the same output

File: sru/cuda_functional.py [hunk 11]
```diff
@@ -4,8 +4,8 @@
 
 from torch.utils.cpp_extension import load
 sources = [
-    os.path.join(os.path.dirname(__file__), "sru_cuda_impl.cpp"),
-    os.path.join(os.path.dirname(__file__), "sru_cuda_kernel.cu"),
+    os.path.join(os.path.dirname(__file__), os.path.join("csrc", "sru_cuda_impl.cpp")),
+    os.path.join(os.path.dirname(__file__), os.path.join("csrc", "sru_cuda_kernel.cu")),
```
Comment (hpasapp) [inline sru/cuda_functional.py:8]: You can actually put these all into a single call to join:

```
os.path.join(os.path.dirname(__file__),, 'csrs', 'sru_cuda_kernel.cu')
```
```

File: sru/ops.py (hunk 4 above)
Comment (taoleicn) [reply sru/ops.py:217]: haha yes. I didn't write a todo list. Moving the test code will be one of them.

File: sru/csrc/main_test_cpp.cpp (hunk 1 above)
Comment (taoleicn) [reply sru/csrc/main_test_cpp.cpp:8]: Yep. It's not easy to do so right now since the code has to be separately compiled thru `cmake` and `make`. Need to save and load a model artifact. I definitely wanna include some tests. Need some help since I'm not an expert (maybe Jing / speech modeling can help).

File: sru/csrc/sru_cpu_impl.cpp (hunk 2 above)
Comment (taoleicn) [reply sru/csrc/sru_cpu_impl.cpp:264]: Will do.

File: sru/modules.py (hunk 5 above)
Comment (taoleicn) [reply sru/modules.py:43]: Gotcha. Will do.

File: sru/__init__.py (hunk 8 above)
Comment (taoleicn) [reply sru/__init__.py:8]: Thanks. I'll use `sru.modules`. The try except code was some legacy code that long time ago I or some people are not using the code as a library. so I added `from sru_functional import *` to support that. 
But it became a problem since many report issues are failures to compile the code and the try-except code hide the real error messages / stack trace information. so i decide to just make it a library-like import.

File: sru/csrc/main_test_cpp.cpp (hunk 9 above)
Comment (taoleicn) [reply sru/csrc/main_test_cpp.cpp:1]: I can do that. I need your help on writing the `CMakeLists.txt` though. Need to add_library(..) from a different directory.
any idea?

File: sru/cuda_functional.py (hunk 11 above)
Comment (taoleicn) [reply sru/cuda_functional.py:8]: oh good to know. Will do.

File: sru/modules.py [hunk 12]
```diff
@@ -3,230 +3,78 @@
...
+        dropout: float, optional
+            the dropout value applied between layers (default=0)
+        rnn_dropout: float, optional
+            [DEPRECATED] the variational dropout value (default=0)
+        bidirectional: bool, optional
+            if True, set the module as a bidirectional SRU (default=False)
+        n_proj: int, optional
+            if non-zero, factorize the ``weight`` parameter matrix as a product of two
+            parameter matrices, using an innder dimension ``n_proj`` (default=0)
+        use_tanh: bool, optional
```
Comment (jshapiro-asapp) [inline sru/modules.py:60]: for the deprecated params (rnn_dropout, use_tanh, v1) is it worth adding a note around the "correct" way to achieve the same behavior? E.g. if I wanted to use tanh activation, is there a different flag I should set instead?

Comment (jshapiro-asapp) [reply sru/modules.py:60]: Also does deprecated imply these params are not torchscript compatible?

File: sru/modules.py [hunk 13]
```diff
@@ -361,22 +210,95 @@ def forward(self, input, c0=None, mask_pad=None, **kwargs):
...
+        # get dropout mask
+        if self.training and (self.dropout > 0):
+            mask_c = self.get_dropout_mask_((batch_size, self.output_size), self.dropout)
+        else:
+            mask_c = None
+
         # compute U, V
         #   U is (length, batch_size, output_size * num_matrices)
         #   V is (output_size*2,) or (length, batch_size, output_size * 2) if provided
+        U, V = self.compute_UV(input, c0, mask_pad)
```
Comment (jshapiro-asapp) [inline sru/modules.py:233]: nit: having U & V be capitalized makes me think they are constants. Possibly worth lowercasing?

File: sru/modules.py [hunk 14]
```diff
@@ -361,22 +210,95 @@ def forward(self, input, c0=None, mask_pad=None, **kwargs):
...
+        # apply elementwise recurrence to get hidden states h and c
+        h, c = self.apply_recurrence(U, V,
+                                     residual, c0,
+                                     scale_val,
+                                     mask_c,
+                                     mask_pad)
+        return h, c
+
+    def apply_recurrence(self,
+                         U: Tensor,
```
Comment (jshapiro-asapp) [inline sru/modules.py:244]: nit: having capitalized variables makes me think they're cosntants. Possibly worth lowercasing?  (though I do see the use in this case, so up to you)

File: sru/ops.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+            hidden_size,
+            k,
+            activation_type,
+            has_skip_term,
+            scale_x.item() if scale_x is not None else 1.0,
+            is_custom
+        )
+
+
+@torch.jit.unused
```
Comment (jshapiro-asapp) [inline sru/ops.py:82]: this is cool, didn't know this decorator existed!

File: sru/modules.py (hunk 13 above)
Comment (hpasapp) [reply sru/modules.py:233]: to me, it makes them look like they are tensors, or matrices. which I think they are?

**Comment by hpasapp** [comment]
Very gentle preference to run `flake8` against the code. Just makes the formatting match up with PEP8 standard formatting. Since we are moving in this direction with ACE (ACE is PEP8/flake8 compliant), maybe now-ish could be a good time to consider using flake8 against sru folder too?

To run flake8:
```
pip install flake8
flake8
```
(there are also plugins available for sublime etc, that show up any flake8 issues in the editor directlry)

File: sru/csrc/main_test_cpp.cpp (hunk 9 above)
Comment (hpasapp) [reply sru/csrc/main_test_cpp.cpp:1]: sure. Can you provide a README.md on how to run cmake please? If I run it from csrs, I get an error about findtorch.cmake not found:

```
(base) ubuntu@hughohp2:~/git/sru/sru/csrc$ cmake .
...
CMake Error at CMakeLists.txt:4 (find_package):
  By not providing "FindTorch.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Torch", but
  CMake did not find one.

  Could not find a package configuration file provided by "Torch" with any of
  the following names:

    TorchConfig.cmake
    torch-config.cmake

  Add the installation prefix of "Torch" to CMAKE_PREFIX_PATH or set
  "Torch_DIR" to a directory containing one of the above files.  If "Torch"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/persist/git/sru/sru/csrc/CMakeFiles/CMakeOutput.log".
```

File: sru/csrc/main_test_cpp.cpp (hunk 1 above)
Comment (hpasapp) [reply sru/csrc/main_test_cpp.cpp:8]: what code needs to be compiled using `make`? dont suppose... could you add a README.md with full build instructions please?

File: sru/modules.py (hunk 12 above)
Comment (hpasapp) [reply sru/modules.py:60]: I tend to agree that it might be useful to have a brief description of why the functionality is being removed altogether, or how to achieve the functionality using a different approach.

File: sru/version.py [hunk 16]
```diff
@@ -1 +1 @@
-__version__ = '2.3.5'
+__version__ = '2.4.0'
```
Comment (hpasapp) [inline sru/version.py:1]: Minor version sounds like the appropriate version number to change to me :) (assuming this change is backwards-compatible? which I think, it is?)

File: sru/csrc/main_test_cpp.cpp (hunk 9 above)
Comment (hpasapp) [reply sru/csrc/main_test_cpp.cpp:1]: Ok, I set TORCH_LIBRARY to `/Users/nycloaner03/conda/lib/python3.7/site-packages/torch/share/cmake/Torch`, and that configured and generated ok.

File: sru/csrc/main_test_cpp.cpp (hunk 1 above)
Comment (hpasapp) [reply sru/csrc/main_test_cpp.cpp:8]: ok, I see. we first run cmake, then run make. forgot about this. so... what I would do is separte the build from the tests, like in the instructions, someone would run:

```
cmake
make
pytest
```

Comment (hpasapp) [reply sru/csrc/main_test_cpp.cpp:8]: like eg this is how it worked fro deepcl: https://github.com/hughperkins/DeepCL/blob/master/doc/Build.md

Screen Shot 2020-08-25 at 11 58 52 AM
Screen Shot 2020-08-25 at 11 59 04 AM

File: sru/modules.py (hunk 13 above)
Comment (jshapiro-asapp) [reply sru/modules.py:233]: I wasn't familiar with that notation, but I see how that could make sense!

File: sru/modules.py (hunk 12 above)
Comment (taoleicn) [reply sru/modules.py:60]: @jshapiro-asapp @hpasapp  I marked them as deprecated because i don't want users to use the options anymore:
(1) the options are for backward compatibility since v1 and won't give better performance;
(2) i want to remove these options and therefore i can clean up the cpp/cu code.

in other words, the "correct" way is not to change the default value and not use these options anymore.
any thoughts?

Comment (hpasapp) [reply sru/modules.py:60]: "tanh is deprecated because minimal performance improvement, and increases codebase size. This option will be removed at the next major version upgrade.
"rnn_dropout is deprecated because minimal performance improvement, and increases codebase size. This option will be removed at the next major version upgrade"

File: test/sru/test_torchscript.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import pytest
+import torch
+import sru
+
+@pytest.mark.parametrize("bidirectional", [False, True])
```
Comment (hpasapp) [inline test/sru/test_torchscript.py:5]: Your using parametrize to test different options is nice :)

File: sru/modules.py [hunk 18]
```diff
@@ -0,0 +1,654 @@
...
+            if c0.dim() != 3:
+                raise ValueError("There must be 3 dimensions for (num_layers, batch_size, output_size)")
+            c0_ = [ x.squeeze(0) for x in c0.chunk(self.num_layers, 0) ]
+
+        if self.input_to_hidden is None:
+            prevx = input
+        else:
+            prevx = self.input_to_hidden(input)
+        lstc = []
+        i = 0
```
Comment (hpasapp) [inline sru/modules.py:599]: I see in one of the commits you reverted using `enumerate` here? I think I prefer the `enumerate` version on the whole, since shorter, easier to read. Is there any reason why the `enumerate` version was reverted?

File: test/test.sh [hunk 19]
```diff
@@ -0,0 +1,10 @@
+set -e
+python test/test_ts_cpp.py > py_out.txt
+cd sru/csrc/
+mkdir build
+cd build
+cmake -DCMAKE_PREFIX_PATH="$(python -c 'import torch; import os.path; print(os.path.join(os.path.dirname(torch.__file__), "share", "cmake"))')" ..
```
Comment (hpasapp) [inline test/test.sh:6]: nice :)

**Comment by hpasapp** [comment]
Adding the flake8 is nice :) Can we add eg a `setup.cfg` to setup the required max line length please? Currently, the scripts give 'line too long' errors in my editor. Looks like you are using maximum line length 100? Can we add please eg `setup.cfg` with following contents?

setup.cfg:
```
[flake8]
max_line_length = 100
```

File: .flake8 [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+max-line-length = 100
+extend-ignore = E203
+exclude =
+    docs
+    DrQA
+    classification
+    language_model
+    misc
+    speech
+    test
```
Comment (hpasapp) [inline .flake8:11]: gentle preference to remove `test` folder from the exclusions.

File: sru/__init__.py [hunk 21]
```diff
@@ -4,8 +4,5 @@
 sru provides a PyTorch implementation of the simple recurrent neural network cell described
 in "Simple Recurrent Units for Highly Parallelizable Recurrence."
 """
-from .version import __version__
-try:
-    from .sru_functional import *
-except:
-    from sru_functional import *
+from sru.version import __version__  # noqa: F401
+from sru.modules import SRU, SRUCell  # noqa: F401
```
Comment (hpasapp) [inline sru/__init__.py:8]: looks like you *might* be able to avoid noqa, if you put after this:

```
__all__ = ['SRU', 'SRUCell']
```

https://stackoverflow.com/a/31079085
